runtime/src/iree-amd-aie/driver/xrt/native_executable.cc

```diff
@@ -0,0 +1,257 @@
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_amd_aie_hal_xrt_native_executable_flatbuffer_verify(
+              executable_params->executable_data));
+
+  iree_amd_aie_hal_xrt_ExecutableDef_table_t executable_def =
+      iree_amd_aie_hal_xrt_ExecutableDef_as_root(
+          executable_params->executable_data.data);
+
+  flatbuffers_string_vec_t entry_points_vec =
+      iree_amd_aie_hal_xrt_ExecutableDef_entry_points_get(executable_def);
+  iree_host_size_t entry_count = flatbuffers_string_vec_len(entry_points_vec);
+
+  flatbuffers_string_t xclbin_fb =
+      iree_amd_aie_hal_xrt_ExecutableDef_xclbin_get(executable_def);
+
+  iree_amd_aie_hal_xrt_AsmInstDef_vec_t asm_instrs_vec =
+      iree_amd_aie_hal_xrt_ExecutableDef_asm_instrs_get(executable_def);
+
+  iree_host_size_t entry_point_count =
+      flatbuffers_string_vec_len(entry_points_vec);
+  // Calculate the total number of characters across all entry point names. This
+  // is only required when tracing so that we can store copies of the names as
+  // the flatbuffer storing the strings may be released while the executable is
+  // still live.
+  iree_host_size_t total_entry_point_name_chars = 0;
+  IREE_TRACE({
+    for (iree_host_size_t i = 0; i < entry_count; i++) {
+      const char* entry_name = flatbuffers_string_vec_at(entry_points_vec, i);
+      total_entry_point_name_chars += flatbuffers_string_len(entry_name);
+    }
+  });
+
+  iree_host_size_t total_size =
+      sizeof(*executable) + entry_count * sizeof(executable->entry_points[0]) +
+      total_entry_point_name_chars;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0,
+      iree_allocator_malloc(host_allocator, total_size, (void**)&executable));
+  IREE_TRACE(
+      char* string_table_buffer =
+          (char*)((char*)executable + sizeof(*executable) +
+                  entry_point_count * sizeof(executable->entry_points[0])));
+
+  iree_hal_resource_initialize(&iree_hal_xrt_native_executable_vtable,
+                               &executable->resource);
+
+  xrt::xclbin xclbin = xrt::xclbin(xclbin_fb);
+  device.register_xclbin(xclbin);
+  xrt::hw_context context(device, xclbin.get_uuid());
+  executable->host_allocator = host_allocator;
+  executable->entry_point_count = entry_point_count;
+  for (iree_host_size_t i = 0; i < entry_point_count; i++) {
+    std::string entry_name = flatbuffers_string_vec_at(entry_points_vec, i);
+    iree_amd_aie_hal_xrt_AsmInstDef_table_t asminst_def =
+        iree_amd_aie_hal_xrt_AsmInstDef_vec_at(asm_instrs_vec, i);
+    flatbuffers_uint32_vec_t asm_inst =
+        iree_amd_aie_hal_xrt_AsmInstDef_asm_inst_get(asminst_def);
+    uint32_t num_instr = flatbuffers_uint32_vec_len(asm_inst);
+    std::unique_ptr<xrt::kernel> kernel;
+    std::unique_ptr<xrt::bo> instr;
+    try {
+      auto kernel = std::make_unique<xrt::kernel>(context, entry_name);
+      auto instr = std::make_unique<xrt::bo>(
+          device, num_instr * sizeof(uint32_t), XCL_BO_FLAGS_CACHEABLE,
```

Comment:
Please document here regarding the `XCL_BO_FLAGS_CACHEABLE` and group id 0.. It's a great source of confusion. We'd want others checking the code to be able to understand why such magic symbols / numbers.